src/encoded/static/libs/bootstrap/tooltip.js

```diff
@@ -0,0 +1,120 @@
+        // Start a timer that might hide the tooltip after a second passes. It won't hide the
+        // tooltip if they're now hovering over the tooltip itself.
+        this.timer = setTimeout(() => {
+            this.timer = null;
+            if (!this.hoveringOverTooltip) {
+                this.setState({ tooltipDisplayed: false });
+            }
+        }, 1000);
+    }
+
+    handleTriggerFocus() {
+        // Show tooltip when keyboard focus lands on tooltip trigger.
+        this.setState({ tooltipDisplayed: true });
+    }
+
+    handleTriggerBlur() {
+        // Hide tooltip when keyboard focus leaves tooltip trigger.
+        this.setState({ tooltipDisplayed: false });
+    }
+
+    handleTooltipMouseEnter() {
+        // Mouse started hovering over the tooltip itself. Prevent the tooltip from hiding even if
+        // the one-second trigger time expires.
+        this.hoveringOverTooltip = true;
+    }
+
+    handleTooltipMouseLeave() {
+        // Mouse stopped hovering over the tooltip itself. If the timer's not running, hide the
+        // tooltip, otherwise let the existing timer hide the tooltip when it expires.
+        this.hoveringOverTooltip = false;
+        if (!this.timer) {
+            this.setState({ tooltipDisplayed: false });
+        }
+    }
+
+    render() {
+        const { trigger, direction, tooltipId, css } = this.props;
+        const tooltipCss = `tooltip ${direction}`;
+        const wrapperCss = `tooltip-container${css ? ` ${css}` : ''}`;
+        return (
+            <div className={wrapperCss}>
+                {this.state.tooltipDisplayed ?
+                    <div className={tooltipCss} role="tooltip" id={tooltipId} onMouseEnter={this.handleTooltipMouseEnter} onMouseLeave={this.handleTooltipMouseLeave}>
+                        <div className="tooltip-arrow" />
+                        <div className="tooltip-inner">{this.props.children}</div>
+                    </div>
+                : null}
+                <button
+                    aria-describedby={this.state.tooltipDisplayed ? tooltipId : ''}
+                    onMouseEnter={this.handleTriggerMouseEnter}
+                    onMouseLeave={this.handleTriggerMouseLeave}
+                    onFocus={this.handleTriggerFocus}
+                    onBlur={this.handleTriggerBlur}
+                    className="tooltip-container__trigger"
+                >
+                    {trigger}
+                </button>
+            </div>
+        );
+    }
+}
+
+Tooltip.propTypes = {
+    trigger: PropTypes.element.isRequired, // Visible tooltip triggering component
+    tooltipId: PropTypes.string.isRequired, // HTML ID of tooltip <div>; unique within page
+    direction: PropTypes.oneOf(['left', 'top', 'bottom', 'right']), // Direction of bootstrap tooltip location
+    css: PropTypes.string, // CSS classes to add to tooltip-container wrapper class
+    children: PropTypes.node,
+};
+
+Tooltip.defaultProps = {
+    direction: 'bottom',
```

Comment:
where are "bottom", "top", "left", "right" used?

Reply:
I don’t understand the question.

Reply:
@forresttanaka I assume direction means placement/position of the tooltip. I assume assume it is changeable by users since you expose it as a props. If these are true, where are alternates to "bottom"-setting used in the code?

Reply:
They’re in tooltips.js PropTypes, which in dev builds also does live value checking. As you see in modals.js, I’m a fan of documenting everything. But the past few months I’ve been burned enough about commenting too much or commenting too little that I’ve just about given up.